File: src-java/src/main/java/kigh/rmt/timer/InMemoryTimerRepository.java
package kigh.rmt.timer;

import org.springframework.stereotype.Component;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

@Component
public class InMemoryTimerRepository implements TimerRepository {
    private final Map<TimerId, Timer> map = new HashMap<>();

    @Override
    public void add(TimerMetadata metadata) {
        if (exists(metadata.id)) {
            throw new IllegalArgumentException("duplicated id");
        }
        Timer timer = TimerFactory.create(metadata);
        map.put(metadata.id, timer);
    }

    @Override
    public boolean exists(TimerId id) {
        return map.containsKey(id);
    }

    @Override
    public Timer get(TimerId id) {
        return null;
    }

    @Override
    public Collection<TimerMetadata> listMetadata() {
        return null;
    }
}

File: src-kotlin/src/main/kotlin/kigh/rmt/timer/InMemoryTimerRepository.kt
package kigh.rmt.timer

import org.springframework.stereotype.Component
import java.lang.IllegalArgumentException
import java.util.*

@Component
class InMemoryTimerRepository : TimerRepository {
    private val map: MutableMap<TimerId, Timer> =
        HashMap()

    override fun add(metadata: TimerMetadata) {
        require(!exists(metadata.id)) { "duplicated id" }
        val timer = TimerFactory.create(metadata)
        map[metadata.id] = timer
    }

    override fun exists(id: TimerId): Boolean {
        return map.containsKey(id)
    }

    override fun get(id: TimerId): Timer {
        return map.get(id) ?: throw IllegalArgumentException()
    }

    override fun listMetadata(): Collection<TimerMetadata> {
        return map.values.map { TimerMetadata(it.id, it.name)}
    }
}